gateway/blocks_backend.go

```diff
@@ -229,12 +263,65 @@ func (bb *BlocksBackend) Head(ctx context.Context, path ImmutablePath) (ContentP
 		return ContentPathMetadata{}, nil, err
 	}
 
-	return md, fileNode, nil
+	sz, err := fileNode.Size()
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	if _, ok := fileNode.(files.Directory); ok {
+		return md, NewHeadResponseForDirectory(sz), nil
+	}
+
+	if _, ok := fileNode.(*files.Symlink); ok {
+		return md, NewHeadResponseForSymlink(sz), nil
+	}
+
+	if f, ok := fileNode.(files.File); ok {
+		return md, NewHeadResponseForFile(f, sz), nil
+	}
+
+	return ContentPathMetadata{}, nil, fmt.Errorf("unsupported UnixFS file type")
 }
 
+// emptyRoot is a CAR root with the empty identity CID. CAR files are recommended
+// to always include a CID in their root, even if it's just the empty CID.
+// https://ipld.io/specs/transport/car/carv1/#number-of-roots
+var emptyRoot = []cid.Cid{cid.MustParse("bafkqaaa")}
+
 func (bb *BlocksBackend) GetCAR(ctx context.Context, p ImmutablePath, params CarParams) (ContentPathMetadata, io.ReadCloser, error) {
 	pathMetadata, err := bb.ResolvePath(ctx, p)
 	if err != nil {
+		rootCid, err := cid.Decode(strings.Split(p.String(), "/")[2])
+		if err != nil {
+			return ContentPathMetadata{}, nil, err
+		}
+
+		var buf bytes.Buffer
+		cw, err := storage.NewWritable(&buf, emptyRoot, car.WriteAsCarV1(true))
```

Comment:
@hacdias @aschmahmann  Could we keep a meaningful root for exports done with `/ipfs/cid` (without any sub-path)?

Reply:
Hmm. The line you highlighted is only for paths that cannot be resolved - the other CARs (line 344) use the CID of the last segment as the root CID. Only if `/ipfs/{cid}` fails to resolve with a `isErrNotFound` is this code path happening. When would that happen?

Reply:
Good call, I wasn't looking too closely when I commented here. We shouldn't be able to hit this codepath with a failed `/ipfs/cid` resolution, and if we did we should be returning an error not writing out an empty CAR file.